src/NuGet.Core/NuGet.Protocol.Core.Types/SourceCacheContext.cs

```diff
@@ -95,18 +79,41 @@ private TimeSpan GetCacheTime(DateTimeOffset? maxAge, TimeSpan defaultTime)
             return timeSpan;
         }
 
-        public string GeneratedTempFolder { get; } = Path.Combine(Path.GetTempPath(), "NuGet", "TempCache", _tempFolderGuid);
+        public string GeneratedTempFolder
+        {
+            get
+            {
+                if (_generatedTempFolder == null)
+                {
+                    var newTempFolder = Path.Combine(
+                        Path.GetTempPath(),
+                        "NuGet",
+                        "TempCache",
+                        Guid.NewGuid().ToString());
+
+                    Interlocked.CompareExchange(ref _generatedTempFolder, newTempFolder, null);
```

Comment:
`Interlocked.CompareExchange` will atomically set `_generatedTempFolder` to `newTempFolder` if `_generatedTempFolder` is null. In other words, it allows lazy initialization as well as thread safety. If two threads enter the `if` at the same time, both will attempt the `CompareExchange` but only one will actually assign its own `newTempFolder`. Both will see that same `newTempFolder` when they `return _generatedTempFolder`.

Concerning why this has lazy initialization, a null `_generatedTempFolder` saves the `Directory.Delete` in the `Dispose` method. I chose to do this optimization because I have a hunch that most users do not specify `-NoCache`, which cases this `GeneratedTempFolder` to be populated with files.